### PR TITLE
fix: `instanceof`

### LINKED ISSUE #220: regression in `rc-6`
Input:
```php
$var = $original instanceof self;
```

I know what i start helping with parser, but some issue is difficult for me and will be great if you help with some issues too.

### PR DESCRIPTION
**Pull request by evilebottnawi**
partial #220, need fix for `static` too

### CODE REVIEW

File: src/parser/variable.js [hunk 1]
```diff
@@ -43,7 +43,11 @@ module.exports = {
     ) {
       result = this.node();
       const name = this.read_namespace_name();
-      if (this.token != this.tok.T_DOUBLE_COLON && this.token != "(") {
+      if (
+        this.token != this.tok.T_DOUBLE_COLON &&
+        this.token != "(" &&
+        ["parentreference", "selfreference"].indexOf(name.kind) === -1
```
Comment (ichiriac) [inline src/parser/variable.js]: it's micro optimisations, but as it's potentially executed many times, it's in principle faster to write this expression like this : 

```
 name.kind !== "parentreference" && 
 name.kind !== "selfreference"
```

**Comment by ichiriac** [comment]
ah, of course, sorry, I wasn't understanding the question. In principle the static is a node so I was guessing that could be the as the others nodes :

```
name.kind !== "staticreference"
````

**Comment by ichiriac** [comment]
give me some minutes, I will try something, I think I understand your problem around static

**Comment by ichiriac** [comment]
First define a debug option in your test header :

```js
const debug = {
  parser: {
    debug: true
  },
  lexer: {
    debug: true
  }
};
```

Next write the failing test, by enabling the debug :
```js
it('instanceof static', () => {
      expect(parser.parseEval('$foo instanceof static;',  debug)).toMatchSnapshot();
    });
```

Next run the test and cheout the debug output, it wil explain how it's parsing your content :
```js
      Error: T_VARIABLE from 1,0         - to 1,4       "$foo"
          at lexer.Object.<anonymous>.lexer.next (...\src\lexer.js:425:15)
          at lexer.next [as lex] (...\src\lexer.js:333:20)
          at parser.lex (...\src\parser.js:575:29)
          at parser.lex [as next] (...\src\parser.js:504:8)
          at parser.next [as parse] (...\src\parser.js:274:8)
          at engine.parse [as parseEval] (...\src\index.js:126:22)
          at Function.parseEval (...\src\index.js:114:15)
          at Object.parseEval (...\test\snapshot\variable.test.js:15:21)
          at Object.asyncFn (...\node_modules\jest-jasmine2\build\jasmine_async.js:82:37)
          at resolve (...\node_modules\jest-jasmine2\build\queue_runner.js:52:12)
    console.log src\parser.js:442
      Line 1 : T_VARIABLE>$foo< @-->at engine.parse [as parseEval] (...\src\index.js:126:22)
    console.error src\lexer.js:440
      Error: T_WHITESPACE       from 1,4         - to 1,5       " "
          at lexer.Object.<anonymous>.lexer.next (...\src\lexer.js:425:15)
          at lexer.next [as lex] (...\src\lexer.js:333:20)
          at parser.lex (...\src\parser.js:575:29)
          at parser.lex [as next] (...\src\parser.js:504:8)
          at parser.next [as read_simple_variable] (...\src\parser\variable.js:318:12)
          at parser.read_simple_variable [as read_reference_variable] (...\src\parser\variable.js:280:23)
          at parser.read_reference_variable [as read_variable] (...\src\parser\variable.js:36:21)
          at parser.read_variable [as read_expr_item] (...\src\parser\expr.js:329:19)
          at parser.read_expr_item [as read_expr] (...\src\parser\expr.js:11:23)
          at parser.read_expr [as read_statement] (...\src\parser\statement.js:366:21)
    console.error src\lexer.js:440
      Error: T_INSTANCEOF       from 1,5         - to 1,15      "instanceof"
          at lexer.Object.<anonymous>.lexer.next (...\src\lexer.js:425:15)
          at lexer.next [as lex] (...\src\lexer.js:343:20)
          at parser.lex (...\src\parser.js:575:29)
          at parser.lex [as next] (...\src\parser.js:504:8)
          at parser.next [as read_simple_variable] (...\src\parser\variable.js:318:12)
          at parser.read_simple_variable [as read_reference_variable] (...\src\parser\variable.js:280:23)
          at parser.read_reference_variable [as read_variable] (...\src\parser\variable.js:36:21)
          at parser.read_variable [as read_expr_item] (...\src\parser\expr.js:329:19)
          at parser.read_expr_item [as read_expr] (...\src\parser\expr.js:11:23)
          at parser.read_expr [as read_statement] (...\src\parser\statement.js:366:21)
    console.log src\parser.js:442
      Line 1 : T_INSTANCEOF>instanceof< @-->at parser.read_simple_variable [as read_reference_variable] (...\src\parser\variable.js:280:23)
    console.error src\lexer.js:440
      Error: T_WHITESPACE       from 1,15        - to 1,16      " "
          at lexer.Object.<anonymous>.lexer.next (...\src\lexer.js:425:15)
          at lexer.next [as lex] (...\src\lexer.js:333:20)
          at parser.lex (...\src\parser.js:575:29)
          at parser.lex [as next] (...\src\parser.js:504:8)
          at parser.next [as read_expr] (...\src\parser\expr.js:67:53)
          at parser.read_expr [as read_statement] (...\src\parser\statement.js:366:21)
          at parser.read_statement [as read_top_statement] (...\src\parser\statement.js:67:21)
          at parser.read_top_statement [as read_start] (...\src\parser\main.js:18:19)
          at parser.read_start [as parse] (...\src\parser.js:276:23)
          at engine.parse [as parseEval] (...\src\index.js:126:22)
    console.error src\lexer.js:440
      Error: T_STATIC   from 1,16        - to 1,22      "static"
          at lexer.Object.<anonymous>.lexer.next (...\src\lexer.js:425:15)
          at lexer.next [as lex] (...\src\lexer.js:343:20)
          at parser.lex (...\src\parser.js:575:29)
          at parser.lex [as next] (...\src\parser.js:504:8)
          at parser.next [as read_expr] (...\src\parser\expr.js:67:53)
          at parser.read_expr [as read_statement] (...\src\parser\statement.js:366:21)
          at parser.read_statement [as read_top_statement] (...\src\parser\statement.js:67:21)
          at parser.read_top_statement [as read_start] (...\src\parser\main.js:18:19)
          at parser.read_start [as parse] (...\src\parser.js:276:23)
          at engine.parse [as parseEval] (...\src\index.js:126:22)
    console.log src\parser.js:442
      Line 1 : T_STATIC>static< @-->at parser.read_expr [as read_statement] (...\src\parser\statement.js:366:21)
    console.error src\lexer.js:440
      Error: ";"        from 1,22        - to 1,23      ";"
          at lexer.Object.<anonymous>.lexer.next (...\src\lexer.js:425:15)
          at lexer.next [as lex] (...\src\lexer.js:333:20)
          at parser.lex (...\src\parser.js:575:29)
          at parser.lex [as next] (...\src\parser.js:504:8)
          at parser.next [as read_expr_item] (...\src\parser\expr.js:315:18)
          at parser.read_expr_item [as read_expr] (...\src\parser\expr.js:11:23)
          at parser.read_expr (...\src\parser\expr.js:67:60)
          at parser.read_expr [as read_statement] (...\src\parser\statement.js:366:21)
          at parser.read_statement [as read_top_statement] (...\src\parser\statement.js:67:21)
          at parser.read_top_statement [as read_start] (...\src\parser\main.js:18:19)
    console.log src\parser.js:442
      Line 1 : ';'>;< @-->at parser.read_expr_item [as read_expr] (...\src\parser\expr.js:11:23)
    console.log src\parser.js:442
      Line 1 : the end of file (EOF)>< @-->at parser.read_expr_item [as read_expr] (...\src\parser\expr.js:11:23)
```

Here what you can see is : for each token, you know its type, T_VARIABLE, and offsets, contents, but also the stack that used to consume it.

Next you have the parser debug, saying the same thing, but from its own scope :  Line 1 : T_VARIABLE>$foo< @-->at engine.parse [as parseEval] (...\src\index.js:126:22) from the the token was consumed.


Here we see that the T_STATIC token was consumed by  Line 1 : T_STATIC>static< @-->at parser.read_expr [as read_statement] (...\src\parser\statement.js:366:21)
read_statement -> read_expr 
https://github.com/glayzzle/php-parser/blob/master/src/parser/statement.js#L366 

And from the stack we can see exactly from where the next statement was launched :
at parser.next [as read_expr] (...\src\parser\expr.js:67:53)
https://github.com/glayzzle/php-parser/blob/master/src/parser/expr.js#L67

So to reproduce the path of the parser :

read_statement 
  -> goes on default case
  -> reads an expression
       --> ok, variable got from read_expr_item
       --> current token if instanceof, lets do a binary node (L67)

And next you can see that we have a parse error : 

```js
SyntaxError: Parse Error : syntax error on line 1

      298 |   message += " on line " + this.lexer.yylloc.first_line;
      299 |   if (!this.suppressErrors) {
    > 300 |     const err = new SyntaxError(
      301 |       message,
      302 |       this.filename,
      303 |       this.lexer.yylloc.first_line

      at parser.Object.<anonymous>.parser.raiseError (src/parser.js:300:17)
      at parser.raiseError [as error] (src/parser.js:345:15)
      at parser.error [as read_expr_item] (src/parser/expr.js:431:12)
      at parser.read_expr_item [as read_expr] (src/parser/expr.js:11:23)
      at parser.read_expr (src/parser/expr.js:67:60)
      at parser.read_expr [as read_statement] (src/parser/statement.js:366:21)
      at parser.read_statement [as read_top_statement] (src/parser/statement.js:67:21)
      at parser.read_top_statement [as read_start] (src/parser/main.js:18:19)
      at parser.read_start [as parse] (src/parser.js:276:23)
      at engine.parse [as parseEval] (src/index.js:126:22)
```

So here you find-out  from where the error was raised : ` at parser.error [as read_expr_item] (src/parser/expr.js:431:12)`

That means in `read_expr_item` function you have not satisfied any parsing path, bad syntax.

I'll continue to how to check syntax with the PHP grammar

**Comment by ichiriac** [comment]
Here the reference file of PHP, that's from where all the code was written and should evolve. If you have any doubt on how it should behave, you can read this :
https://github.com/php/php-src/blob/master/Zend/zend_language_parser.y

In our cas wee just need to know how PHP handles T_STATIC tokens, so do a search in this file, you will find something like :

https://github.com/php/php-src/blob/master/Zend/zend_language_parser.y#L86
This is the way it will associate with other nodes

https://github.com/php/php-src/blob/master/Zend/zend_language_parser.y#L177
The token representation as text (for errors messages for example)

```
semi_reserved:
	  reserved_non_modifiers
	| T_STATIC | T_ABSTRACT | T_FINAL | T_PRIVATE | T_PROTECTED | T_PUBLIC
;
```
It's a SEMI-RESEVED keywords, so when `semi_reserved` is used, `T_STATIC` is also included

...etc...

We found nothing, so lets digg with the instanceof token T_INSTANCEOF :

```
expr: ...
expr T_INSTANCEOF class_name_reference
...

class_name_reference:
		class_name		{ $$ = $1; }
	|	new_variable	{ $$ = $1; }
;

class_name:
		T_STATIC
			{ zval zv; ZVAL_INTERNED_STR(&zv, ZSTR_KNOWN(ZEND_STR_STATIC));
			  $$ = zend_ast_create_zval_ex(&zv, ZEND_NAME_NOT_FQ); }
	|	name { $$ = $1; }
;

namespace_name:
		T_STRING								{ $$ = $1; }
	|	namespace_name T_NS_SEPARATOR T_STRING	{ $$ = zend_ast_append_str($1, $3); }
;
name:
		namespace_name								{ $$ = $1; $$->attr = ZEND_NAME_NOT_FQ; }
	|	T_NAMESPACE T_NS_SEPARATOR namespace_name	{ $$ = $3; $$->attr = ZEND_NAME_RELATIVE; }
	|	T_NS_SEPARATOR namespace_name				{ $$ = $2; $$->attr = ZEND_NAME_FQ; }
;
```

So we've got it !

In order to write it down here, we should use the same names with a read prefix :

```
read_class_name_reference: () => {
 if (t_string, t_token, T_NS_SEPARATOR) return read_namespace;
 if (t_static) return new node('staticreference');
 // and here it's tricky !
}
```

As you can read, you also have a new_variable definition, you can read it here :
```
new_variable:
		simple_variable
			{ $$ = zend_ast_create(ZEND_AST_VAR, $1); }
	|	new_variable '[' optional_expr ']'
			{ $$ = zend_ast_create(ZEND_AST_DIM, $1, $3); }
	|	new_variable '{' expr '}'
			{ $$ = zend_ast_create(ZEND_AST_DIM, $1, $3); }
	|	new_variable T_OBJECT_OPERATOR property_name
			{ $$ = zend_ast_create(ZEND_AST_PROP, $1, $3); }
	|	class_name T_PAAMAYIM_NEKUDOTAYIM simple_variable
			{ $$ = zend_ast_create(ZEND_AST_STATIC_PROP, $1, $3); }
	|	new_variable T_PAAMAYIM_NEKUDOTAYIM simple_variable
			{ $$ = zend_ast_create(ZEND_AST_STATIC_PROP, $1, $3); }
;
```

It's meaning you can a variable and use chaining you can find this on read_variable, but the trcky part is that it's also includes : `class_name T_PAAMAYIM_NEKUDOTAYIM simple_variable`

It's like resolving some factorisation : `class_name or (class_name + '::' + variable)` you have to check after if it's followed by `T_PAAMAYIM_NEKUDOTAYIM` + a simple variable, if it's a variable it's ok, else error

**Comment by ichiriac** [comment]
BTW, I still use `T_DOUBLE_COLON` because at the time I was converting the code, T_PAAMAYIM_NEKUDOTAYIM` was removed, here more information about this confusion : https://philsturgeon.uk/php/2013/09/09/t-paamayim-nekudotayim-v-sanity/

**Comment by ichiriac** [comment]
I let you try to figure out how to change the syntax, ask me if you don't find something

**Comment by evilebottnawi** [comment]
@ichiriac a lot of information for me :smile: I don't know `PHP` so deep :smile: Maybe you can specify which lines need to be corrected? And then I will investigate your decision and in the future I will be able to better correct the errors.

**Comment by ichiriac** [comment]
Ok, better is to accept the PR and continue the fix by referencing the issue

**Pull request merged**